### DISCUSSION TITLE
Updating another table after crud operation in a table.

### DISCUSSION THREAD

**Discussion by Knightshrestha** [pocketbase/pocketbase#6122] [Q&A] [answered]
I have tables for series list and book list with 1:n relation. book list has foreign key,

I managed to make the update after create and update work but the delete has errors.

Code
```
/// <reference path="../pb_data/types.d.ts" />
onRecordAfterDeleteSuccess((event) => {
	const record = event.record;

	const series_id = record.get('source_series');

	const lastBookInSeries = $app.findRecordsByFilter(
		'remote_book_info',
		'source_series = {:series_id}',
		'-order',
		0,
		0,
		{ series_id: record.get('source_series'), visible: true }
	);

	$app.logger().info(
		'Delete Book: Update Series with old book',
		'series_id',
		series_id,
		'length',
		lastBookInSeries.length,
		'latest',
		lastBookInSeries[0]
	);

	event.next();
}, 'remote_book_info');
```
Error
```
ERROR DELETE /api/collections/pbc_3015858974/records/e01u33n1301rbix
└─ Failed to delete record. Make sure that the record is not part of a required relation reference.
   └─ sql: transaction has already been committed or rolled back; failed query: SELECT `remote_book_info`.* FROM `remote_book_info` WHERE [[remote_book_info.source_series]] = {:t7OY8z} ORDER BY [[remote_book_info.order]] DESC
```
By removing the codes, i found the error occurs when I am fetching all records associated to the relevant series, what is happening?
Despite it saying failed to delete record, the record actually gets deleted.

Working code for create hook
```
/// <reference path="../pb_data/types.d.ts" />
onRecordAfterCreateSuccess((event) => {
	const record = event.record;

	if (!record) {
		event.next();
	}

	const series_id = record.get('source_series');
	const current_order = record.get('order') || 1;

	const lastBookInSeries = $app.findRecordsByFilter(
		'remote_book_info',
		'source_series = {:series_id}',
		'-order',
		0,
		0,
		{ series_id, visible: true }
	);

	const s = lastBookInSeries.map((i) => ({
		order: i.get('order'),
		id: i.id,
	}));

	const compare = s[0].order < current_order;

	if (!compare) event.next();

	const series = $app.findRecordById('remote_series_info', series_id);
	series.set('latest_book_id', record.id);
	series.set('latest_book_cover', record.get('cover_image'));
	series.set('book_count', lastBookInSeries.length);

	$app.save(series);

	$app.logger().info(
		'Create Book, Update Series with new book',
		'book_id',
		record.id,
		'file_name',
		record.get('cover_image')
	);
	event.next();
}, 'remote_book_info');

```

**Comment by ganigeorgiev** [comment]
[`onRecordAfterDeleteSuccess`](https://pocketbase.io/docs/js-event-hooks/#onrecordafterdeletesuccess) is triggered AFTER the delete operation has been committed so even if you throw an error the record will be still deleted.

The returned error is most likely happening because the delete operation is internally executed in a transaction. Try replacing `$app` with `event.app`.

The longer explanation is: because `event.app` could be changed by the user and could point to a transactional app instance (_aka. similar to the txApp in https://pocketbase.io/docs/js-records/#transaction_) in the JSVM there is an extra `$app` overwrite that assigns it to `event.app` in attempt to prevent deadlocks but this seems to not work specifically with the `*After*` db hooks because the event app instance is reverted to the original app after the transaction completes but `$app` will still point to the old one (_the instance with the already committed transaction_). I'll try to experiment a little bit later and eventually will consider adding a special condition with the next release for the `*After*` hooks for consistency in the JSVM.

**Reply by ganigeorgiev** [reply]
_Couple things as a side-note:_

- There shouldn't be situation where `event.record` is null in `onRecordAfterCreateSuccess` hook.
- the `visible: true` param doesn't seem to be binded to the filter (_if the value must be always true you can simply add `&& visible = true` in the filter_)
- the `if` statements that wraps the `event.next()` seems to be missing `return`, otherwise the code after it will still execute

**Reply by Knightshrestha** [reply]
```
	try {
		const lastBookInSeries = event.app.findRecordsByFilter(
			'remote_book_info',
			'source_series = {:series_id}',
			'-order',
			0,
			0,
			{ series_id }
		);

		console.log("event.app", lastBookInSeries);
	} catch (error) {
		console.error('event.app', error);
	}

		try {
			const lastBookInSeries = $app.findRecordsByFilter(
				'remote_book_info',
				'source_series = {:series_id}',
				'-order',
				0,
				0,
				{ series_id }
			);

			console.log('app', lastBookInSeries);
		} catch (error) {
			console.error('app', error);
		}
```
Errors
```
2024/12/16 14:57:44 event.app GoError: sql: transaction has already been committed or rolled back; failed query: SELECT `remote_book_info`.* FROM `remote_book_info` WHERE [[remote_book_info.source_series]] = {:tzgPya} ORDER BY [[remote_book_info.order]] DESC
2024/12/16 14:57:44 app GoError: sql: transaction has already been committed or rolled back; failed query: SELECT `remote_book_info`.* FROM `remote_book_info` WHERE [[remote_book_info.source_series]] = {:tq9alv} ORDER BY [[remote_book_info.order]] DESC
```
It didn't changed anything.

**Reply by ganigeorgiev** [reply, marked as answer]
Hm, in this case it could be a bug with the hook. I'll investigate it a little bit later today. As a temp workaround specifically for the delete, you can use the regular `onRecordDelete` hook or the related request `onRecordDeleteRequest` hook:

```js
onRecordDelete((event) => {
        event.next(); // perform the delete (the below operation should execute as part of the delete transaction)

	...
}, 'remote_book_info');
```

**Reply by Knightshrestha** [reply]
This worked. Thank you.

**Reply by ganigeorgiev** [reply]
This ended up a little more complicated then I initially thought and the cause of it was incorrect syncing between the base `OnModel*` and the proxy `OnRecord*` hook events. I've submitted a hotfix in [v0.23.8 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.8) and now the `onRecordAfterDeleteSuccess` should work as well.

Further optimizations in relation to this will be available with the next major v0.24.0 release, including significant speed up for the record cascade delete operations.

**Comment by muety** [comment]
Just came across the exact same issue. Amazing to see that it has been fixed literally _today_!

**Reply by patronovski** [reply]
this guy is really amazing with the support he is giving
